### DISCUSSION TITLE
How to install jdtls?

### DISCUSSION THREAD

**Discussion by JRS-Developer** [mason-org/mason.nvim#337] [Q&A] [answered]
Hello, I have tried to install jdtls using MasonInstall and it says that it's installed successfully but when I open a java file it shows up this message: "Client 1 quit with exit code 1 and signal 0"

I have already installed OpenJDK 17 using APT with Ubuntu 16.04.

I tried executing jdtls from ~/.local/share/nvim/mason/bin  and it shows up this message:

Traceback (most recent call last):
  File "./jdtls", line 17, in <module>
    jdtls.main(sys.argv[1:])
  File "/home/jose/.local/share/nvim/mason/packages/jdtls/bin/jdtls.py", line 68, in main
    parser.add_argument("--validate-java-version", default=True, action=argparse.BooleanOptionalAction)
AttributeError: module 'argparse' has no attribute 'BooleanOptionalAction'

**Comment by williamboman** [comment, marked as answer]
Hello! So the `jdtls` executable requires `python3` to be installed (`which python3` should produce a path). Not only that, they require Python >= 3.9 to be installed. It's a bit unfortunate if you ask me because it hurts portability, but it is what it is. What Python version do you have installed (`python3 --version`)?

**Reply by smjonas** [reply]
For anyone else reading this, be careful! Don't blindly upgrade your Python version on Linux. I followed [this](https://phoenixnap.com/kb/upgrade-python) tutorial to upgrade to 3.9. However, this broke many things on my machine, including the update manager. I had to revert it. Not sure if there is a way to upgrade such that `jdtls` points to a different python version than the one that `python3 --version` returns. As of now, this requirement prevents me from using jdtls...

**Reply by mikehaertl** [reply]
@smjonas You could try [pyenv](https://github.com/pyenv/pyenv). It allows you to 

* have multiple versions of python installed in parallel
* select the python version to use on your user shell
* select the python version to use on the system level (change not recommended!)
* select the python version to use for a specific project/directory (requires pyenv-virtualenv plugin)

While the docs may look intimidating the setup is actually quite easy and only requires some lines in your `.bashrc` and `.profile` (plus installation of some OS dependent build tools to install other python versions).

Disclaimer: I'm not a python expert and also did not try if this fixes the problem as my system already uses python 3.10 as default. But it's worth a try.
